sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/EliminateUnnecessaryJoin.scala

```diff
@@ -0,0 +1,91 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.adaptive
+
+import org.apache.spark.sql.catalyst.planning.ExtractSingleColumnNullAwareAntiJoin
+import org.apache.spark.sql.catalyst.plans.{Inner, LeftAnti, LeftSemi}
+import org.apache.spark.sql.catalyst.plans.logical.{Join, LocalRelation, LogicalPlan}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.execution.joins.HashedRelationWithAllNullKeys
+
+/**
+ * This optimization rule detects and eliminates unnecessary Join:
+ * 1. Join is single column NULL-aware anti join (NAAJ), and broadcasted [[HashedRelation]]
+ *    is [[HashedRelationWithAllNullKeys]]. Eliminate join to an empty [[LocalRelation]].
+ *
+ * 2. Join is inner join, and either side of join is empty. Eliminate join to an empty
+ *    [[LocalRelation]].
+ *
+ * 3. Join is left semi join
+ *    3.1. Join right side is empty. Eliminate join to an empty [[LocalRelation]].
+ *    3.2. Join right side is non-empty and condition is empty. Eliminate join to its left side.
+ *
+ * 4. Join is left anti join
+ *    4.1. Join right side is empty. Eliminate join to its left side.
+ *    4.2. Join right side is non-empty and condition is empty. Eliminate join to an empty
+ *         [[LocalRelation]].
+ *
+ * This applies to all joins (sort merge join, shuffled hash join, broadcast hash join, and
+ * broadcast nested loop join), because sort merge join and shuffled hash join will be changed
+ * to broadcast hash join with AQE at the first place.
+ */
+object EliminateUnnecessaryJoin extends Rule[LogicalPlan] {
+
+  private def isRelationWithAllNullKeys(plan: LogicalPlan) = plan match {
+    case LogicalQueryStage(_, stage: BroadcastQueryStageExec)
+      if stage.resultOption.get().isDefined =>
+      stage.broadcast.relationFuture.get().value == HashedRelationWithAllNullKeys
+    case _ => false
+  }
+
+  private def checkRowCount(plan: LogicalPlan, hasRow: Boolean): Boolean = plan match {
```

Comment:
This is added compared to previous `EliminateJoinToEmptyRelation.scala`.

Reply:
nit comment: we need the `hasRow` param? How about simply checking `count == 0` like this?
```
private def hasNoRow(plan: LogicalPlan): Boolean = plan match {
    case LogicalQueryStage(_, stage: QueryStageExec) if stage.resultOption.get().isDefined =>
      stage.getRuntimeStatistics.rowCount match {
        case Some(count) => count == 0
        case _ => false
      }
    case _ => false
  }
```
Then,
```
    case j @ Join(_, _, Inner, _, _) if hasNoRow(j.left) ||
    ...
    case j @ Join(_, _, LeftSemi, condition, _) =>
      if (checkRowCount(j.right, hasRow = false)) {
        LocalRelation(j.output, data = Seq.empty, isStreaming = j.isStreaming)
      } else if (condition.isEmpty && !hasNoRow(j.right)) {
```

Reply:
@maropu - sorry I don't quite get here, `checkRowCount(j.right, hasRow = false` still exists in your example. Am I missing anything? Thanks.

Reply:
ah, my bad. I meant `checkRowCount` -> `hasNoRow `:
```
    case j @ Join(_, _, Inner, _, _) if hasNoRow(j.left) ||
    ...
    case j @ Join(_, _, LeftSemi, condition, _) =>
      if (hasNoRow(j.right)) {
        LocalRelation(j.output, data = Seq.empty, isStreaming = j.isStreaming)
      } else if (condition.isEmpty && !hasNoRow(j.right)) {
```

Reply:
@maropu - I thought about this too, but I think this is dangerous as we want to make sure the child output is known, and row count is larger than zero.

There can be the case where `hasNoRow()` returns false but the plan is not executed in AQE at all (not `LogicalQueryStage`) or result is not known yet.

Reply:
Ah, okay. I got your point. sgtm